### ISSUE TITLE
Unpublished queries show up in search results for other users

### ISSUE THREAD

**Issue by arikfr** [getredash/redash#3217]
Looks like that in one of the changes to the queries API, we started showing unpublished queries to all the users. This is not the intended behavior, although I'm no longer sure if we should revert back.

We have the option to allow editing queries/dashboards by other users in a feature flag, that soon will become enabled for all. If we keep the old logic (hide unpublished queries from non owners) it will make it really hard to load the queries a user show see if we also need to filter unpublished queries.

Would love to hear opinions.

**Comment by kravets-levko** [comment]
I think with new permissions settings for queries/dashboards neither approach is correct. We should show items that user has access to - own items, published items or items with granted access. It will be the most correct behavior from the user's point of view.

**Comment by kravets-levko** [comment]
Okay, got it :thinking:

**Comment by susodapop** [comment]
I think this change should be reverted to make application behavior consistent. It seems to me that the whole justification for a "Published Queries" mechanic was privacy. Don't publish a query until you're ready for others to see it. This is how Dashboard publishing still works. But with this behavior removed, why would a user choose to publish a query? Or phrased another way, why **wouldn't** a user want to publish a query? If it's to allow the query on dashboards, then publishing seems like an irrelevant extra step. If it's to allow embedding the visualization outside of Redash then "publish" seems like the wrong verb. It could probably just be a checkbox instead.

Perhaps a compromise on the first question above would be to change the UI for Publishing. Suppose a user wants to add a visualization from an unpublished query to their dashboard. The query would appear in the "Add Widget" modal search on the dashboard screen. And when they select the query it would pop a message box explaining "This query hasn't been published. Would you like to publish it now?". If they choose later to unpublish, they can visit the query screen and toggle the button.

**Comment by justinclift** [comment]
> Don't publish a query until you're ready for others to see it.

On that note, noticed something interesting while attempting to replicate the documented behaviour today.

With a new Query, if you change the title from the default text string (at least in v7), the query automatically changes to Published.  It's possible to Unpublish it afterwards, but it did catch me out initially.

**Comment by susodapop** [comment]
> On that note, noticed something interesting while attempting to replicate the documented behaviour today.

What you describe is the documented behavior. 

> To publish a query, give it a name or click the `Publish` button.

**Comment by justinclift** [comment]
Oh.  I guess that just seems weird to me.

Literally the first thing I do with any new query is name it something appropriate, to match it's purpose.

It doesn't make sense (to me yet) for it's name to have any bearing on whether I'd want it Published or not.  Shouldn't they be two separate things?

**Comment by susodapop** [comment]
> Oh. I guess that just seems **really** weird to me.

I agree it's non-obvious. A typical user isn't going to assume that changing the name of their query has side-effects.

**Comment by arikfr** [comment]
In the past a query was considered as a draft when it was named "New Query". Later on we added the explicit the "is_draft" state (False=published, True=unpublished). To keep things simple to use and consistent with historic behavior we added the option to auto publish when naming a query (it's a flag you can disable).

**Comment by justinclift** [comment]
Interesting. :smile:

**Comment by ranbena** [comment]
> > Oh. I guess that just seems **really** weird to me.
> 
> I agree it's non-obvious. A typical user isn't going to assume that changing the name of their query has side-effects.

Understanding the gravity of the publish feature, I think we should revoke auto-publish-on-rename asap @arikfr.

**Comment by arikfr** [comment]
@ranbena let's keep things in perspective: 

1. It was like this for a long period of time and no one seemed to care.
2. When we first released the publish workflow there was a backlash from users complaining about the extra steps needed... This is when we introduced the old behavior of "name to publish".

We can (and should) explore alternatives to the current workflow, but no need to do anything in a rush :)

**Comment by ranbena** [comment]
> We can (and should) explore alternatives to the current workflow, but no need to do anything in a rush :)

Understood and legit - here's a quick followup https://github.com/getredash/redash/pull/3998

**Comment by justinclift** [comment]
Personally, I just reckon we should make sure the docs match the actual behaviour as they're out-of-sync at the moment.  Caused me confusion trying to get things working "as per the docs".  Other people may have been caught out, no idea.

Anyway, Jesse's patch to do that (when it passes review) sounds like the right thing to sync them up. :smile:
